### PR TITLE
[TextField] Fix `focused` prop so it triggers focus state

### LINKED ISSUE #985: [TextField] Focused prop is not setting focus on the element
# Issue summary

The `focused` prop on the `<TextField />` component is not setting focus on the input.

## Expected behavior

When `focused` is set to `true`, the input (or textarea if `multiline` is set) should be on its focus state.

## Actual behavior

The input/textarea is on its default state.

## Steps to reproduce the problem

1. Create a `<TextField />` with the `focused` prop set to `true`
2. Verify the input is not on its focus state

## Reduced test case

https://codesandbox.io/s/n41rrjovl0?module=App.js

## Specifications

- Are you using the React components? (Y/N): Yes
- Polaris version number: 3.6.0
- Browser: Google Chrome 71.0.3578.98
- Device: MacBook Pro
- Operating System: macOS Mojave 10.14.2

### PR DESCRIPTION
**Pull request by amireynoso**
### WHY are these changes introduced?

Resolves #985 

The <TextField /> component accepts a `focused` prop which should force the field to be focused. However, it is not currently working (take a look at [the issue](https://github.com/Shopify/polaris-react/issues/985) for a reduced test case).

### WHAT is this pull request doing?

I based my changes in the [`<SearchField />`](https://github.com/Shopify/polaris-react/tree/master/src/components/TopBar/components/SearchField) implementation (component and tests), which also accepts a `focused` prop and it works as expected. This is a breakdown of the changes:

- It now takes the `focused` prop into consideration when setting the value for `state.focus` (used to be `false` by default)
- Forces the input's focus state on `componentDidMount()` if the prop is `true`.
- Forces  the input's focus state (or blur state) on `componentDidUpdate()` in case the prop is toggled
- Scoped the tests related to this prop into their own `describe` block
- The current test claimed to be testing something it was not testing (nor doing): "focuses input **and calls onFocus()** when focused prop has been updated to true"
   - I refactored this test a little bit to test that the element is in focus when the prop is set to `true`
- Added two new tests to check the prop can be toggled

### How to 🎩

🖥 [Local development instructions](https://github.com/Shopify/polaris-react/blob/master/README.md#development)
🗒 [General tophatting guidelines](https://github.com/Shopify/polaris-react/blob/master/documentation/Tophatting.md)
📄 [Changelog guidelines](https://github.com/Shopify/polaris-react/blob/master/documentation/Versioning%20and%20changelog.md)

<details>
<summary>Copy-paste this code in <code>playground/Playground.tsx</code>:</summary>

```jsx
import * as React from 'react';
import {Page, TextField} from '../src';

interface State {
  value: string;
}

export default class Playground extends React.Component<{}, State> {
  state = {
    value: 'Lorem ipsum',
  };

  handleChange = (value: string) => {
    this.setState({value});
  };

  render() {
    return (
      <Page title="Playground">
        <TextField
          label="Focused text field"
          value={this.state.value}
          onChange={this.handleChange}
          multiline
          focused
        />
      </Page>
    );
  }
}
```

</details>

### 🎩 checklist

* [x] Tested on [mobile](https://github.com/Shopify/polaris-react/blob/master/documentation/Tophatting.md#cross-browser-testing)
* [x] Tested on [multiple browsers](https://help.shopify.com/en/manual/intro-to-shopify/shopify-admin/supported-browsers)
* [x] Tested for [accessibility](https://github.com/Shopify/polaris-react/blob/master/documentation/Accessibility%20testing.md)

### CODE REVIEW

File: src/components/TextField/TextField.tsx [hunk 1]
```diff
@@ -132,21 +132,29 @@ class TextField extends React.PureComponent<CombinedProps, State> {
 
     this.state = {
       height: null,
-      focus: false,
+      focus: props.focused || false,
```
Comment (AndrewMusgrave) [inline src/components/TextField/TextField.tsx]: We can probably leave this as
```suggestion
      focus: props.focused,
```

File: src/components/TextField/TextField.tsx [hunk 2]
```diff
@@ -132,21 +132,29 @@ class TextField extends React.PureComponent<CombinedProps, State> {
...
   }
 
-  componentDidUpdate({focused}: CombinedProps) {
-    if (
-      this.input &&
-      focused !== this.props.focused &&
-      this.props.focused === true
-    ) {
+  componentDidMount() {
+    if (this.input && this.props.focused) {
```
Comment (AndrewMusgrave) [inline src/components/TextField/TextField.tsx]: The ref should always be available in `componentDidMount`? We should be able to omit that check. The JS styleguide also prefers early returns so we can probably change this to
```tsx
if (!this.props.focused) return;
this.input.focus();
```

File: src/components/TextField/TextField.tsx [hunk 3]
```diff
@@ -132,21 +132,29 @@ class TextField extends React.PureComponent<CombinedProps, State> {
...
       this.input.focus();
     }
   }
 
+  componentDidUpdate({focused: wasFocused}: CombinedProps) {
+    const {focused} = this.props;
+
+    if (this.input) {
```
Comment (AndrewMusgrave) [inline src/components/TextField/TextField.tsx]: Same comment about the early return here as well

File: src/components/TextField/TextField.tsx [hunk 4]
```diff
@@ -132,21 +132,29 @@ class TextField extends React.PureComponent<CombinedProps, State> {
...
+      if (!wasFocused && focused) {
+        this.input.focus();
+      } else if (wasFocused && !focused) {
+        this.input.blur();
```
Comment (AndrewMusgrave) [inline src/components/TextField/TextField.tsx]: Should we blur when the props change? Using the `focused` prop forces focus on the textfield but omitting the prop doesn't necessarily mean they want the field to be blurred. What are your thoughts on this?

File: src/components/TextField/TextField.tsx (hunk 1 above)
Comment (amireynoso) [reply src/components/TextField/TextField.tsx]: True! I just did this for type "sanity" but I'll change it back.

File: src/components/TextField/TextField.tsx (hunk 2 above)
Comment (amireynoso) [reply src/components/TextField/TextField.tsx]: Sounds good!

File: src/components/TextField/TextField.tsx (hunk 4 above)
Comment (amireynoso) [reply src/components/TextField/TextField.tsx]: In this case, the `blur` even would only trigger if the `focused` prop changed values and it wouldn't be triggered on mount. 

One use case is if `focused` is being set programatically based on an interaction with another component. If the focus is removed (programatically), I'd expect it to trigger the blur event.

I saw this on the [`<SearchField />`](https://github.com/Shopify/polaris-react/tree/master/src/components/TopBar/components/SearchField) component and thought it was a nice addition.

File: src/components/TextField/TextField.tsx (hunk 1 above)
Comment (amireynoso) [reply src/components/TextField/TextField.tsx]: Oh, nevermind. Since `focused` is possibly `undefined`, we need to set a `boolean` default value.

**Comment by amireynoso** [comment]
Hey @AndrewMusgrave, thanks for your review! I addressed the comments, would you mind taking another look?

**Comment by amireynoso** [comment]
@solonaarmstrong Hey Solona! Would you help me with a review of this PR?  I wasn't sure who else to tag, feel free to tag someone else. Thank you!

**Comment by AndrewMusgrave** [comment]
Sorry I already gave feedback so this doesn't show up on my review list anymore, I'll get to it later today 🙏

File: src/components/TextField/TextField.tsx (hunk 1 above)
Comment (AndrewMusgrave) [reply src/components/TextField/TextField.tsx]: Sorry I thought it optional 😅

File: UNRELEASED.md [hunk 5]
```diff
@@ -12,6 +12,8 @@ Use [the changelog guidelines](https://git.io/polaris-changelog-guidelines) to f
 
 ### Bug fixes
 
+- Fixed the `focused` prop on `TextField` so it sets the focus state ([#985](https://github.com/Shopify/polaris-react/pull/985))
```
Comment (AndrewMusgrave) [inline UNRELEASED.md]:
```suggestion
- Fixed the `focused` prop on `TextField` so it sets the focus state ([#990](https://github.com/Shopify/polaris-react/pull/990))
```

File: src/components/TextField/TextField.tsx (hunk 4 above)
Comment (AndrewMusgrave) [reply src/components/TextField/TextField.tsx]: Works for me, how about you @solonaarmstrong ?

**Comment by solonaarmstrong** [comment]
What's the difference between `focused` and `autoFocus`. It looks as though `autoFocus` is the one that actually places the cursor in the field?

**Comment by AndrewMusgrave** [comment]
`Autofocus` will focus the textfield when it renders, focused maintains focus on the textfield

File: UNRELEASED.md (hunk 5 above)
Comment (amireynoso) [reply UNRELEASED.md]: 🤦‍♀️Thanks for catching that!

**Pull request merged**